### PR TITLE
Fix bug in LoadProperty for complex and complex collection

### LINKED ISSUE #2471: DataServiceContext.LoadProperty fails to load collection of complex types.
*DataServiceContext.LoadProperty fails to load collection of complex types..*

### Assemblies affected

*OData .Net lib 7.x*

### Reproduce steps

*Run attached sample project.*

### Expected result

*Count of addresses should be printed.*

### Actual result

*Exception occurs: 
System.InvalidCastException
  HResult=0x80004002
  Message=Unable to cast object of type 'Trippin.Location' to type 'System.Collections.ObjectModel.Collection`1[Trippin.Location]'.
  Source=<Cannot evaluate the exception source>
  StackTrace:
<Cannot evaluate the exception stack trace>

  This exception was originally thrown at this call stack:
    Microsoft.OData.Client.Metadata.ClientPropertyAnnotation.SetValue(object, object, string, bool)
    Microsoft.OData.Client.Materialization.ODataLoadNavigationPropertyMaterializer.ReadImplementation()
    Microsoft.OData.Client.Materialization.ODataMaterializer.Read()
    Microsoft.OData.Client.MaterializeAtom.MoveNextInternal()
    Microsoft.OData.Client.MaterializeAtom.MoveNext()
    Microsoft.OData.Client.LoadPropertyResult.ReadPropertyFromAtom(Microsoft.OData.Client.Metadata.ClientPropertyAnnotation)
    Microsoft.OData.Client.LoadPropertyResult.LoadProperty()
    Microsoft.OData.Client.DataServiceContext.LoadProperty(object, string, System.Uri)
    Microsoft.OData.Client.DataServiceContext.LoadProperty(object, string)
    LoadProperty.Program.Main(string[]) in Program.cs*
[LoadProperty.zip](https://github.com/OData/odata.net/files/9211857/LoadProperty.zip)

### PR DESCRIPTION
**Pull request by gathogojr**
### Issues

*This pull request fixes #2471.*

### Description

We rely on `ODataLoadNavigationPropertyMaterializer` to materialize load property result for complex and complex collection property. Complex and complex collection properties are structural properties, the reliance on `ODataLoadNavigationPropertyMaterializer` is by design - we use `ODataReaderEntityMaterializer` as well to materialize complex and complex collection properties, since just like entities, they're returned as `ODataResource`.
The bug was in how we apply the materialized values to the respective properties on the client type instance.
Due to a missing condition check, the complex collections were being treated as if they are a single entity value.

I fixed the bug by applying the missing check. By doing so, [this block in `LoadPropertyResult` class](https://github.com/OData/odata.net/blob/4e263ca4b94d01f2d105d8282c62ec272cfcd1bb/src/Microsoft.OData.Client/LoadPropertyResult.cs#L220) does the job of applying the complex collection to the property - after first clearing the existing items.

### Checklist (Uncheck if it is not completed)

- [ ] *Test cases added*
- [ ] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

File: src/Microsoft.OData.Client/Materialization/ODataLoadNavigationPropertyMaterializer.cs [hunk 1]
```diff
@@ -99,12 +102,11 @@ protected override bool ReadImplementation()
                         this.MaterializeEntryPlan,
                         this.responseInfo.IsContinuation);
                 }
-                else
+                else if (property.EdmProperty.Type.TypeKind() == EdmTypeKind.Entity)
```
Comment (habbes) [inline src/Microsoft.OData.Client/Materialization/ODataLoadNavigationPropertyMaterializer.cs:105]: In which block do complex properties and/or complex collection properties get handled?

Comment (gathogojr) [reply src/Microsoft.OData.Client/Materialization/ODataLoadNavigationPropertyMaterializer.cs:105]: @habbes This block https://github.com/OData/odata.net/blob/4e263ca4b94d01f2d105d8282c62ec272cfcd1bb/src/Microsoft.OData.Client/LoadPropertyResult.cs#L220

**Review by KenitoInc — APPROVED** [review]

File: test/FunctionalTests/Tests/DataServices/UnitTests/Client.TDD.Tests/Microsoft.OData.Client.TDDUnitTests.csproj [hunk 2]
```diff
@@ -81,6 +81,7 @@
     <Compile Include="Tests\Materialization\CollectionValueMaterializationPolicyTests.cs" />
     <Compile Include="Tests\Materialization\EntryValueMaterializationPolicyUnitTests.cs" />
     <Compile Include="Tests\Materialization\CamelCasedTypeMaterializationTests.cs" />
+    <Compile Include="Tests\LoadPropertyTests.cs" />
```
Comment (xuzhg) [inline test/FunctionalTests/Tests/DataServices/UnitTests/Client.TDD.Tests/Microsoft.OData.Client.TDDUnitTests.csproj:84]: shall we put the new test file under "Materialization" subfolder?

**Review by xuzhg — APPROVED** [review]

File: test/FunctionalTests/Tests/DataServices/UnitTests/Client.TDD.Tests/Microsoft.OData.Client.TDDUnitTests.csproj (hunk 2 above)
Comment (gathogojr) [reply test/FunctionalTests/Tests/DataServices/UnitTests/Client.TDD.Tests/Microsoft.OData.Client.TDDUnitTests.csproj]: @xuzhg Done

**Review by Nthemba — APPROVED** [review]

**Review by KenitoInc — APPROVED** [review]

**Pull request merged**